resolver/resolve.go

```diff
@@ -358,6 +358,10 @@ resolveLoop:
 				// Defensive: This should normally not happen.
 				continue
 			}
+			// Check if we got NXDomain and whether we should try another resolver.
+			if rrCache.IsNXDomain() && tryAll {
+				continue
+			}
```

Comment:
We should retry at SERVFAIL but also on NXDOMAIN? NXDOMAIN should be very clear in it's meaning. This check may also cause upstream block detection to fail (and may even cause the whole upstream block detection to be useless with `blockedIf=empty` if we always try a different resolver)

Reply:
The important tidbit is the `tryAll` which is only true in special cases. See `scopes.go`.